File: 200302016_Kenan_Imamovic_homeowork_1.py
import json

questions = json.loads(open("Quiz_Questions.json").read())["questions"]

player = {
    "name": "",
    "currentQuestionIndex": 0,
    "correctAnswers": 0,
    "incorrectAnswers": 0,
}

def main():
    global player
    player["name"] = input('What is your name? (Type "exit" to exit) ')
    
    if player["name"].lower() == "exit":
        return
    else:
        print(f'\nWelcome to my quiz, {player["name"]}!\n')
        game()

def game():
    global player
    if player["currentQuestionIndex"] < len(questions):
        ask_question(player["currentQuestionIndex"])
    else:
        generate_report()

def ask_question(index):
    global player
    question_data = questions[index]
    question = question_data["question"]
    choices = question_data["choices"]

    print(f'\n{question}\n')
    for i, choice in enumerate(choices):
        print(f"{i + 1}) {choice}")

    user_answer = input(f"\nEnter your choice (1-{len(choices)}): ")
    
    if not validate_answer(user_answer, len(choices)):
        print(f"Invalid input. Please enter a number between 1 and {len(choices)}.")
        ask_question(index)
    else:
        validate_question(question_data, user_answer)

def validate_answer(answer, num_choices):
    return answer.isdigit() and 1 <= int(answer) <= num_choices

def validate_question(question, user_answer):
    global player
    correct_answer = question["correctAnswer"]

    if correct_answer == user_answer:
        player["correctAnswers"] += 1
        print("Correct!")
    else:
        player["incorrectAnswers"] += 1
        print("Incorrect!")

    player["currentQuestionIndex"] += 1
    game()

def generate_report():
    global player
    print("\nQuiz Summary:")
    print(f"Name: {player['name']}")
    print(f"Total Questions: {len(questions)}")
    print(f"Correct Answers: {player['correctAnswers']}")
    print(f"Incorrect Answers: {player['incorrectAnswers']}")

if __name__ == "__main__":
    main()

File: 200302016_Kenan_Imamovic_homeowork_1.cs

// make sure to install Newtonsoft.Json
// command:
// dotnet add package Newtonsoft.Json

﻿using System;
using System.Collections.Generic;
using System.IO;
using Newtonsoft.Json;

class Program
{
    private static Dictionary<string, object> player = new Dictionary<string, object>()
    {
        {"name", ""},
        {"currentQuestionIndex", 0},
        {"correctAnswers", 0},
        {"incorrectAnswers", 0}
    };

    private static List<Dictionary<string, object>> questions;

    static void Main(string[] args)
    {
        questions = JsonConvert.DeserializeObject<Dictionary<string, List<Dictionary<string, object>>>>(File.ReadAllText("Quiz_Questions.json"))["questions"];
        player["name"] = AskForName();

        if (player["name"].ToString().ToLower() == "exit")
        {
            return;
        }
        else
        {
            Console.WriteLine($"\nWelcome to my quiz, {player["name"]}!\n");
            Game();
        }
    }

    static string AskForName()
    {
        Console.Write("What is your name? (Type \"exit\" to exit) ");
        return Console.ReadLine();
    }

    static void Game()
    {
        if ((int)player["currentQuestionIndex"] < questions.Count)
        {
            AskQuestion((int)player["currentQuestionIndex"]);
        }
        else
        {
            GenerateReport();
        }
    }

static void AskQuestion(int index)
{
    var questionData = questions[index];
    var question = questionData["question"].ToString();
    var choices = ((Newtonsoft.Json.Linq.JArray)questionData["choices"]).ToObject<List<string>>();

    Console.WriteLine($"\n{question}\n");

    for (int i = 0; i < choices.Count; i++)
    {
        Console.WriteLine($"{i + 1}) {choices[i]}");
    }

    Console.Write($"\nEnter your choice (1-{choices.Count}): ");
    var userAnswer = Console.ReadLine();

    if (!ValidateAnswer(userAnswer, choices.Count))
    {
        Console.WriteLine($"Invalid input. Please enter a number between 1 and {choices.Count}.");
        AskQuestion(index);
    }
    else
    {
        ValidateQuestion(questionData, userAnswer);
    }
}

static bool ValidateAnswer(string answer, int numberOfChoices)
{
    int parsedAnswer;
    return int.TryParse(answer, out parsedAnswer) && parsedAnswer >= 1 && parsedAnswer <= numberOfChoices;
}


    static void ValidateQuestion(Dictionary<string, object> question, string userAnswer)
    {
        var correctAnswer = question["correctAnswer"].ToString();

        if (correctAnswer == userAnswer)
        {
            player["correctAnswers"] = (int)player["correctAnswers"] + 1;
            Console.WriteLine("Correct!");
        }
        else
        {
            player["incorrectAnswers"] = (int)player["incorrectAnswers"] + 1;
            Console.WriteLine("Incorrect!");
        }

        player["currentQuestionIndex"] = (int)player["currentQuestionIndex"] + 1;
        Game();
    }

    static void GenerateReport()
    {
        Console.WriteLine("\nQuiz Summary:");
        Console.WriteLine($"Name: {player["name"]}");
        Console.WriteLine($"Total Questions: {questions.Count}");
        Console.WriteLine($"Correct Answers: {player["correctAnswers"]}");
        Console.WriteLine($"Incorrect Answers: {player["incorrectAnswers"]}");
    }
}
